### ISSUE TITLE
Invalid use of pre in Blocks.Logical.TriggeredTrapezoid

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#790]
**Reported by stefanv on 16 Aug 2012 20:59 UTC**
We were investigating an issue when we came across this Modelica library component that has a pre around a continuous variable in its definition. This pre(y) is inside a when-equation and y is declared as Blocks.Interfaces.RealOutput y

```
when {initial(),u,not u} then
    endValue = if u then offset + amplitude else offset;
    rate = if u and (rising > 0) then amplitude/rising 
           else if not u and (falling > 0) then -amplitude/falling
           else 0;
    T = if u and not (rising > 0) or not u and not (falling > 0) or not abs(amplitude) > 0 or initial() then time 
        else time + (endValue - pre(y))/rate;
end when;
```

The initial equation section in the source code of this component (in Blocks/Logical.mo) contains this comment:

```
  initial equation
    /* A start value of y is set, because pre(y) is present
     to avoid a warning message from the compiler. However,
     this setting does not have an effect, because y is initialized
     correctly, before pre(y) is used */
  pre(y) = 0;
```

Is this valid?  Judging by the comment, it's a hack to get around some tool's compiler.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/790

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 2 Oct 2012 09:45 UTC**
While looking at this model, which I agree is invalid, and looking at where it is used I found there is a duplicate class in Modelica.Fluid.Examples.AST_BatchPlant.BaseClasses.TriggeredTrapezoid. The duplicate is never used as the class Modelica.Fluid.Examples.AST_BatchPlant.BatchPlant_StandardWater makes use of the original from Modelica.Blocks.

**Comment by modelica-trac-importer** [comment]
**Changelog removed by pharman on 2 Oct 2012 09:45 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 2 Oct 2012 13:04 UTC**
Peter, that should probably be filed as a separate ticket.

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 2 Oct 2012 15:15 UTC**
Good point, see #852

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 2 Oct 2012 15:24 UTC**
I filed this before the 76th design meeting where we adopted the compliance initiative, but since illegal Modelica to get around a bug in a tool is a compliance issue, I'm raising this to a blocker.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Oct 2012 14:38 UTC**
I don't know what tool generates the warning, but it is common to add superfluous initializations in e.g. C-code to pass static code checkers.

That is normally considered good software engineering practice, even if a human can analyze the code and determine that the initialization is not needed in the particular case.

In this case we have:
```
   y = if time < T then endValue - (T - time)*rate else  endValue;
   when {initial(),u,not u} then
       T = if u and not (rising > 0) or not u and not (falling
         > 0) or not abs(amplitude) > 0 or initial() then time else time
         + (endValue - pre(y))/rate;
    end when;
```
Here T seems to use pre(y) - but looking more closely we see that during initialization we will get T=time, regardless of pre(y), and then compute y directly from T.

Written in e.g. C this would be:
```
  int i;
  double y=0; /* Value is not used, but needed to avoid warning from lint */
  double T;
  for(i=0;i<10;++i) {
    if (i==0 || f1(i)) {
      T=(f2(i) || (i==0)) ? i : f3(y, i);
    }
    y=f4(T);
  }
```
Removing the initialization of 'y' would be silly.

Conclusion: The initialization value for y should be kept as it is(\*). However, the comment might be rewritten if found offensive, and as above explain why pre(y) is not used. There are also some improvements needed for the spec list below:
\--
It is not clear what "invalid" refers to in this ticket, but I as far I understand the issue is the use of "pre(y)" in an initial equation (i.e. how y is initialized). That seems like an omission in the standard: the expressions in the "initial equation"-sections should be seen as discrete expressions. 

The time is certainly not changing; and I see no issue with:
```
initial algorithm 
   discreteVar:=f(time); 
```
which is also used in MSL.

*: The alternative to having initial equation pre(y)=0; would be a start-value for y. However, I am not certain that it is specified that pre(y) is initialized to the start-value for continuous-time variables.

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 4 Oct 2012 14:46 UTC**
I too do not know which tool it is that gives a warning (I didn't write the comment in the original code). However, the issue here is not entirely that y is being initialized in order to avoid a warning, but the fact that pre() is used on y. One can't use pre() on a continuous variable, which y is.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Oct 2012 14:54 UTC**
Replying to [comment:6 stefanv]:
> I too do not know which tool it is that gives a warning (I didn't write the comment in the original code). However, the issue here is not entirely that y is being initialized in order to avoid a warning, but the fact that pre() is used on y. One can't use pre() on a continuous variable, which y is.

You can in some cases: The requirement is that pre() is applied to a discrete-time expression which is also a variable. Thus pre(y) for a continuous-time variable is allowed (and frequently used) inside when-clauses, since they are then seen as discrete-time expressions; and it seems that expressions in initial equations also should be seen as discrete-time expressions (in both cases time is fixed).

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 4 Oct 2012 14:59 UTC**
Replying to [comment:7 hansolsson]:
> it seems that expressions in initial equations also should be seen as discrete-time expressions

Which in the specification they aren't, so the model is invalid. So the conclusion can't be that it is kept as it is, either the model must be changed or the specification must be.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Oct 2012 15:23 UTC**
Replying to [comment:8 pharman]:
> Replying to [comment:7 hansolsson]:
> > it seems that expressions in initial equations also should be seen as discrete-time expressions
> 
> Which in the specification they aren't, so the model is invalid. So the conclusion can't be that it is kept as it is, either the model must be changed or the specification must be.

My conclusion above was that the spec should be improved; but the model code should not be changed. Currently the spec requires that pre(y) is unknown during initialization, but there is no way to write initialization equations for pre(y).

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 4 Oct 2012 15:42 UTC**
Replying to [comment:9 hansolsson]:
> there is no way to write initialization equations for pre(y). 

I'd say there is:

```
  when initial() then
    pre(y) = 0;
  end when;
```

which in my view makes matters even worse because it is different behavior between different methods of initializing variables. I've therefore created a ticket #853.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Oct 2012 15:47 UTC**
Replying to [comment:10 pharman]:
> Replying to [comment:9 hansolsson]:
> > there is no way to write initialization equations for pre(y). 
> 
> I'd say there is:
> 
> ```
>   when initial() then
>     pre(y) = 0;
>   end when;
> ```
> 
> which in my view makes matters even worse because it is different behavior between different methods of initializing variables. I've therefore created a ticket #853.

The left-hand side in a when-equation may not be pre(y), see 8.3.5.2.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Oct 2012 20:26 UTC**
As one of the principal designers of the hybrid features (including pre(..)), I would like to point out that the hybrid features have been designed so that pre(..) can be applied in a when-clause and in initial equation/algorithm sections on any type of variable, including non-discrete time variables. The meaning is:

  A variable like y refers to the right limit y(t+eps), eps->0 and pre(y) refers to the left limit y(t-eps), eps->0.
  If event iteration takes place, pre(y) refers to the value from the last event iteration.

If this feature would not be present, several useful models could not be defined.

If this is not stated in this form in the specification, the specification must be corrected (I did not yet have time to check this).

Now, back to the example. During initialization variables der(x1) and pre(x2) are treated as algebraic unknown variables that are independent of x1 and x2. The single assignment rule also holds at initialization and this means if there are n variables we need exactly n equations. For the TriggeredTrapezoid this means there MUST be an equation for pre(y), otherwise the number of unknowns (y, pre(y), endValue, rate, T) and the number of equations does not match.

I removed the initialization of pre(y) in the initial equation section and tried in Dymola: With default options there is a warning "The initial conditions are not fully specified" and Dymola assigns a value of 0.0 to pre(y). In pedantic mode, Dymola triggers an error (since the initial conditions are not fully specified). So, Dymola makes a perfect job.

Note, the alternative would be to set a start value for y:

```
RealOutput y(final start=0.0)

```
since this has exactly the same effect (providing the equation pre(y)=0.0 for initialization). However, this is quite visible to the outside for a setting that does not influence the result (every start value will give exactly the same time behavior of y). Therefore, it is better to set this inside an initial equation section.

To summarize, the model is correct. However, the comment is completely misleading. I removed the comment in 61aa793a203f51ff9a62c617b1cf999056da6ac6 and consider this ticket as fixed (an unclear specification should be a different ticket).
